chips/sam4l/src/adc.rs

```diff
@@ -273,32 +293,61 @@ impl hil::adc::Adc for Adc {
                 // as the CPU clock
                 pm::enable_clock(Clock::PBA(PBAClock::ADCIFE));
                 nvic::enable(nvic::NvicIdx::ADCIFE);
-
-                // turn on GCLK10. The Generic Clock is needed for the ADC to
-                // work, although I have no idea what it is needed for. Its
-                // speed does not seem to matter, so let's just peg it to the
-                // CPU clock
-                scif::generic_clock_enable(scif::GenericClock::GCLK10, scif::ClockSource::CLK_CPU);
-
-                // determine clock divider
-                // we need the ADC_CLK to be a maximum of 1.5 MHz in frequency,
-                // so we need to find the PRESCAL value that will make this
-                // happen.
-                // Formula: f(ADC_CLK) = f(CLK_CPU)/2^(N+2) <= 1.5 MHz
-                // and we solve for N
-                // becomes: N <= ceil(log_2(f(CLK_CPU)/1500000)) - 2
-                let cpu_frequency = pm::get_system_frequency();
-                let divisor = (cpu_frequency + (1500000 - 1)) / 1500000; // ceiling of division
-                clock_divisor = math::log_base_two(math::closest_power_of_two(divisor)) - 2;
-                clock_divisor = cmp::min(cmp::max(clock_divisor, 0), 7); // keep in bounds
-                self.adc_clk_freq.set(cpu_frequency / (1 << (clock_divisor + 2)));
+                // the maximum sampling frequency with the RC clocks is 1/32th of their clock
+                // frequency. This is because of the minimum PRESCAL by a factor of 4 and the
+                // 7+1 cycles needed for conversion in continuous mode. Hence, 4*(7+1)=32.
+                if frequency <= 115000 / 32 {
```

Comment:
Can you go over the math in more detail @brghena ? I am having trouble getting these numbers.